### DISCUSSION TITLE
GitHub(automerge): Take the `bypass_pull_request_allowances ` setting in branch protection into account

### DISCUSSION THREAD

**Discussion by Shegox** [renovatebot/renovate#19548] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.70.5/7d6a4815c5d637affffcddff25cd96d3ed33dd9b

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Renovate doesn't notice that it is allowed to bypass "required reviews" branch protection rule like this one: 
<img width="762" alt="image" src="https://user-images.githubusercontent.com/16856448/209304322-cc91f167-34ba-4cf5-9572-764cebbb6024.png">

I propose that we add an extra check to determine if the Renovate user or App is explicatly (aka ignore teams) allowed to bypass the pull_request requirement
https://github.com/renovatebot/renovate/blob/0fd539266dd3f154373aa6b79fa94a7e38b3f906/lib/modules/platform/github/index.ts#L586-L594

This information is already available within the `bypass_pull_request_allowances`:
```js
{
  url: "https://api.github.com/repos/tobias-something-something/renovate-automerge-branch-protection/branches/main/protection/required_pull_request_reviews",
  dismiss_stale_reviews: false,
  require_code_owner_reviews: false,
  require_last_push_approval: false,
  required_approving_review_count: 1,
  bypass_pull_request_allowances: {
    users: [
      {
        login: "Shegox",
        id: 16856448,
        node_id: "MDQ6VXNlcjE2ODU2NDQ4",
        avatar_url: "https://avatars.githubusercontent.com/u/16856448?v=4",
        gravatar_id: "",
        url: "https://api.github.com/users/Shegox",
        html_url: "https://github.com/Shegox",
        followers_url: "https://api.github.com/users/Shegox/followers",
        following_url: "https://api.github.com/users/Shegox/following{/other_user}",
        gists_url: "https://api.github.com/users/Shegox/gists{/gist_id}",
        starred_url: "https://api.github.com/users/Shegox/starred{/owner}{/repo}",
        subscriptions_url: "https://api.github.com/users/Shegox/subscriptions",
        organizations_url: "https://api.github.com/users/Shegox/orgs",
        repos_url: "https://api.github.com/users/Shegox/repos",
        events_url: "https://api.github.com/users/Shegox/events{/privacy}",
        received_events_url: "https://api.github.com/users/Shegox/received_events",
        type: "User",
        site_admin: false,
      },
    ],
    teams: [
    ],
    apps: [
    ],
  },
}
```

---

I'm classifying this as a bug, because there is a secondary thing which cause the behavior to become flaky and unpredicatable.

The actual "loading" of the branch protection rules happens in this code when `getRepoForceRebase()` is called and is only executed _sometimes_: 
https://github.com/renovatebot/renovate/blob/d6d9ee719d5800b62b83aa2a78f856b26561f0b0/lib/workers/repository/update/branch/reuse.ts#L60-L62

which performs the actual loading of the branchProtectionRules and adds them to the **repository global config** via this: 
https://github.com/renovatebot/renovate/blob/0fd539266dd3f154373aa6b79fa94a7e38b3f906/lib/modules/platform/github/index.ts#L584-L594

and that is only called during the renovation of each branch, when automerge is false. Effectively this causes that if the first branch should be auto merged it works (because then the branch protection rules are not loaded), but if only the second branch should be automerged it won't work because Renovate has loaded the branch protection rules, sees that an approval is missing (which it doesn't really need) and don't perform the automerge. 

Inside the reproduction repository that causes that I can automerge the patch-level node update without problems, while just automerging the `renovate/renovate` container image is not possilbe.

---

Reproduction repository: https://github.com/tobias-something-something/renovate-automerge-branch-protection

---

I can raise a PR for this, but most likely only after January 9th, when I'm back from vacation. 

### Relevant debug logs

<details><summary>Logs</summary>

Logs with only automerge for renovate/renovate; doesn't perform automerge

```
...
DEBUG: Found branch protection (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
DEBUG: Branch protection: PR Reviews are required before merging (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
...
DEBUG: Branch protection: Attempting to merge PR when PR reviews are enabled (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/renovate-renovate-34.x)
       "prNo": 3
DEBUG: Branch protection: Cannot automerge PR until there is an approving review (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/renovate-renovate-34.x)

```

Logs with automerge for node (patch-level) and renovate/renovate, automerges node

```
...
DEBUG: Automerging #2 with strategy auto (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
DEBUG: mergePr(2, renovate/node-14.x) (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
DEBUG: mergePr (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
       "options": {"body": {"merge_method": "rebase"}},
       "url": "repos/tobias-something-something/renovate-automerge-branch-protection/pulls/2/merge"
DEBUG: PR merged (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
       "automergeResult": {
         "sha": "12a4d9be7ff348be7552649149f1320133b88527",
         "merged": true,
         "message": "Pull Request successfully merged"
       },
       "pr": 2
 INFO: PR automerged (repository=tobias-something-something/renovate-automerge-branch-protection, branch=renovate/node-14.x)
       "pr": 2,
       "prTitle": "Update Node.js to v14.21.2"
...

```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by viceice** [comment]
This isn't a bug. Please re-raise as feature request.

**Reply by Shegox** [reply]
I would argue, and that's because why I opened it as a bug report, because it _sometimes_ work that is at least a bit buggy and unexpected for an user. 

E.g. 
In my reproduction repository following automerge works: 
```json
    {
      "automerge": true,
      "matchPackageNames": ["node"],
      "matchUpdateTypes": ["patch"],
      "ignoreTests": true
    }
```

while following doesn't:
```json
    {
      "automerge": true,
      "matchPackageNames": ["renovate/renovate"],
      "ignoreTests": true
    }
```

And this only depends on the internal order Renovate processes the branches (because for the first it doesn't load the branchProtection rules and tries automerging) and for the second it does load the branchProtectionRules (and then doesn't try an automerge).

**Comment by viceice** [comment]
- related #7077

**Comment by viceice** [comment]
@rarkins We had the idea to simply not check branch protections and try to auto-merge. Should we refactor to do that now?

**Reply by rarkins** [reply]
I think so. We can ignore branch protections and:
- log a warning along lines of "Not allowed to automerge"
- save this state for the run and not attempt other automerges if so
- potentially we can query the branch protection status *after* if we want to be helpful, but that is optional

**Reply by viceice** [reply]
can you create a feature request and close the other issue?

**Reply by rarkins** [reply]
#19549
